llvm/lib/Target/AArch64/AArch64FrameLowering.cpp

```diff
@@ -3267,10 +3267,13 @@ bool AArch64FrameLowering::spillCalleeSavedRegisters(
         InsertSEH(MIB, TII, MachineInstr::FrameSetup);
     } else { // The code when the pair of ZReg is not present
       MachineInstrBuilder MIB = BuildMI(MBB, MI, DL, TII.get(StrOpc));
-      if (!MRI.isReserved(Reg1))
+      const AArch64RegisterInfo *RegInfo =
+          static_cast<const AArch64RegisterInfo *>(
+              MF.getSubtarget().getRegisterInfo());
+      if (!RegInfo->isStrictlyReservedReg(MF, Reg1))
```

Comment:
I'm not confident this does what you want: even if you get the liveness right inside FrameLowering, target-independent code like LivePhysRegs will ignore registers marked isReserved().  So I'm not sure any liveness computations based on a reserved LR can be trusted.

Reply:
Maybe not, but I'm not sure what other options we have. It's worked for so far except for this case with the outliner.

Reply:
We could unreserve LR after register allocation, I guess?  That would have roughly the right semantics.

Or we could make the outliner use some other method to try to figure out the "liveness" of LR if it's a reserved register.  I mean, if LR is reserved, the only way it can be live is if the prologue doesn't save/restore it (in which case it's live across the whole function).

Reply:
> We could unreserve LR after register allocation, I guess? That would have roughly the right semantics.

Having `getReservedRegs(MF)` return different things depending on where it's running seems worse than the bug itself.

With shrink-wrapping the prolog can be in successor blocks, so LR can be live on some CFG paths and not others.

Reply:
> Having getReservedRegs(MF) return different things depending on where it's running seems worse than the bug itself.

I think this can happen anyway?  For example, with the frame/base pointer.

> With shrink-wrapping the prolog can be in successor blocks, so LR can be live on some CFG paths and not others.

Oh, right.

Other possibilities:

- We could mess with the allocation orders so LR isn't allocated, despite being an allocatable register.  This has basically all the properties we want... but we don't have the necessary infrastructure to make it straightforward.
- We could introduce a distinction between "reserved for regalloc" and "tracked for liveness".  So we'd have registers where we track liveness, but the register allocator doesn't allocate them.

Maybe worth bringing up on Discourse to see if there are other opinions.

(I'm not trying to make things complicated here, but I really don't want to take shortcuts when it comes to core MIR datastructures.)

Reply:
> I think this can happen anyway? For example, with the frame/base pointer.

Indeed :(

> We could mess with the allocation orders so LR isn't allocated, despite being an allocatable register. This has basically all the properties we want... but we don't have the necessary infrastructure to make it straightforward.

Yeah, I thought about something along these lines. Perhaps if we could dynamically alter regclasses.

I think I'll give the changing reservedness approach a go here since there's already some precedent for it.